service/api_logic/recommendation_logic.py

```diff
@@ -0,0 +1,289 @@
+import random
+from Levenshtein import ratio
+from datetime import datetime, timedelta
+import pandas as pd
+from dto.api_input import UserInteractionSchema
+from dto.api_output import OutputRecommendationList
+from exept.exeptions import NoUsersInDBError, EmptyRecommendationListForUserError
+from logger.logger import Logger
+
+SEC_PER_DAY = 24 * 60 * 60
+MAX_LIMIT_FOR_REC_IN_DAYS = 21
+SCORE_FOR_NOT_PREFER_SPORT = 0.25
+SCORE_FOR_PREFER_SPORT = 0.20
+SCORE_FOR_NOTHING = 0.1
+SCORE_FOR_PREFER_TEAM = 0.2
+BASE_LIMIT_OF_NEWS_FOR_RECS = 5
+MIN_PROMOTION_COF = 1
+MAX_PROMOTION_COF = 4
+
+
+class RecommendationService:
+    def __init__(self, recommendation_dal, user_dal, news_dal):
+        self._recommendation_dal = recommendation_dal
+        self._user_dal = user_dal
+        self._news_dal = news_dal
+        self.logger = Logger("logger", "all.log").logger
+        self._recommendations_list = []
+
+
+    async def auto_hybrid_recommendation_system(self, top_n=BASE_LIMIT_OF_NEWS_FOR_RECS):
+        users_that_want_recs = self._recommendation_dal.get_all_users()
+        if not users_that_want_recs:
+            raise NoUsersInDBError
+
+        interaction_df = self.__get_interactions(users_that_want_recs)
+        users_interaction_matrix = self.__create_interaction_matrix(interaction_df)
+        for user in users_that_want_recs:
+            try:
+                recommendations_for_user_with_score = await self.__user_based_recommendations(user.user_id, users_interaction_matrix, top_n=top_n)
+                if not recommendations_for_user_with_score or not any(recommendations_for_user_with_score):
+                    raise EmptyRecommendationListForUserError(user.user_id)
```

Comment:
We need to talk about it on meeting today.

Concerns:
1) All users query
2) Many requests
3) Bad method name
4) Do we need an exception here?